### PR TITLE
ES6 Migration

### PR DESCRIPTION
**Pull request by noirbizarre**
This PR start ES6 migration with models.

### CODE REVIEW

File: js/components/badges/modal.vue [hunk 1]
```diff
@@ -45,7 +45,8 @@
 <script>
 'use strict';
 
-var API = require('api');
+var API = require('api'),
```
Comment (davidbgk) [inline js/components/badges/modal.vue]: Why not `const` or `let`?

File: js/config.js [hunk 2]
```diff
@@ -1,15 +1,14 @@
-define(['jquery'], function($) {
-    'use strict';
+'use strict';
```
Comment (davidbgk) [inline js/config.js]: Does it make it global?

File: js/components/badges/modal.vue (hunk 1 above)
Comment (noirbizarre) [inline js/components/badges/modal.vue]: Because this file has not been ported to ES6 yet :)

File: js/config.js (hunk 2 above)
Comment (noirbizarre) [inline js/config.js]: In fact with babel-loader, it's now useless.
I will remove it ASAP

File: js/components/badges/modal.vue (hunk 1 above)
Comment (davidbgk) [inline js/components/badges/modal.vue]: Fair enough, there are a lot of other places in that PR where we can get rid of `var`s though ;)

File: js/models/badges.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                        var namespace = name + 's',
+                            operation = 'available_' + name + '_badges';
+
+                        if (!API.hasOwnProperty(namespace) || !API[namespace].hasOwnProperty(operation)) {
+                            throw new BadgeError(`Badge for ${name} does not exists`);
+                        }
+
+                        let badges = this._badges[name] = {};
+
+                        API[namespace][operation]({}, (response) => {
```
Comment (davidbgk) [inline js/models/badges.js]: I think that you can remove brackets and even parenthesis around the response parameter.

File: js/models/base.js [hunk 4]
```diff
@@ -1,26 +1,133 @@
...
+/**
+ * A property getter resolving dot-notation
+ * @param  {Object} obj  The root object to fetch property on
+ * @param  {String} name The optionnaly dotted property name to fetch
+ * @return {Object}      The resolved property value
+ */
+function getattr(obj, name) {
+    if (!obj || !name) return;
+    var names = name.split(".");
+    while(names.length && (obj = obj[names.shift()]));
```
Comment (davidbgk) [inline js/models/base.js]: Interesting, I didn't know that syntax!

File: js/models/base.js [hunk 5]
```diff
@@ -29,90 +136,290 @@ define(['api', 'models/validator', 'vue'], function(API, validator, Vue) {
...
+        }
+        this.$emit('updated', this);
+        this.loading = false;
+    }
+
+    /**
+     * Get an item given its ID
+     */
+    by_id(id) {
+        var filtered = this.items.filter(function(item) {
```
Comment (davidbgk) [inline js/models/base.js]: No arrow function here?

File: js/models/base.js [hunk 6]
```diff
@@ -29,90 +136,290 @@ define(['api', 'models/validator', 'vue'], function(API, validator, Vue) {
...
+        if (!this.query.sort) {
+            return;
+        }
+        return this.query.sort[0] == '-';
+    }
+
+    get has_search() {
+        var op = API[this.$options.ns].operations[this.$options.fetch];
+
+        return op.parameters.filter(function(p) {
```
Comment (davidbgk) [inline js/models/base.js]: No arrow? I wonder if the use of `Boolean` is more appropriated/explicit here.

File: js/models/base.js [hunk 7]
```diff
@@ -29,90 +136,290 @@ define(['api', 'models/validator', 'vue'], function(API, validator, Vue) {
...
+     * @return {[type]}         [description]
+     */
+    fetch(options) {
+        this.query = Object.assign(this.query, options || {});
+        this.loading = true;
+        this.$api(this.$options.ns + '.' + this.$options.fetch, this.query, this.on_fetched);
+        return this;
+    }
+
+    next() {
```
Comment (davidbgk) [inline js/models/base.js]: It smells like [an iterator](http://www.2ality.com/2015/02/es6-iteration.html), is it?

File: js/models/base.js [hunk 8]
```diff
@@ -29,90 +136,290 @@ define(['api', 'models/validator', 'vue'], function(API, validator, Vue) {
...
+
+    sort(field, reversed) {
+        this.sorted = field;
+        this.reversed = reversed !== undefined ? reversed : !this.reversed;
+        this.page = 1;
+        this.data = this.build_page();
+    }
+
+    search(query) {
+        console.log('search', query);
```
Comment (davidbgk) [inline js/models/base.js]: `NotImplementedError` :)

File: js/models/base.js [hunk 9]
```diff
@@ -29,90 +136,290 @@ define(['api', 'models/validator', 'vue'], function(API, validator, Vue) {
...
+                    valB = getattr(b, this.sorted);
+
+                if (valA > valB) {
+                    return this.reversed ? -1 : 1;
+                } else if (valA < valB) {
+                    return this.reversed ? 1 : -1;
+                } else {
+                    return 0;
+                }
+            }.bind(this))
```
Comment (davidbgk) [inline js/models/base.js]: An arrow function will avoid binding `this`.

File: js/models/organization.js [hunk 10]
```diff
@@ -1,70 +1,75 @@
...
+    create() {
+        this.$api('organizations.create_organization', {
+            payload: JSON.stringify(this.$data)
+        }, this.on_fetched);
+    }
+
+    role_for(obj) {
+        var user_id = obj.hasOwnProperty('id') ? obj.id : obj,
+            members = this.members.filter(function(member) {
+                return member.user.id == user_id;
```
Comment (davidbgk) [inline js/models/organization.js]: `===` is safer, everywhere.

File: js/models/organization.js [hunk 11]
```diff
@@ -1,70 +1,75 @@
...
+    }
 
-    return Organization;
-});
+    refuse_membership(request, comment, callback) {
+        this.$api('organizations.refuse_membership', {
+            org: this.id,
+            id: request.id,
+            payload: JSON.stringify({comment: comment})
+        }, function(response) {
```
Comment (davidbgk) [inline js/models/organization.js]: Arrow function? We can maybe turn the API from callbacks to promises but that can be done later.

File: js/models/site.js [hunk 12]
```diff
@@ -1,15 +1,13 @@
...
-    return new Site().fetch();
-});
+export class Site extends Model {
+    fetch() {
+        this.$api('site.get_site', {}, this.on_fetched);
+        return this;
+    }
+};
+
+export var site = new Site().fetch();
```
Comment (davidbgk) [inline js/models/site.js]: I'm not fond of this pattern.

**Comment by davidbgk** [comment]
Terrific work! :+1:

File: js/models/base.js (hunk 4 above)
Comment (yohanboniface) [inline js/models/base.js]: Nice one liner! :)

File: js/models/badges.js (hunk 3 above)
Comment (noirbizarre) [inline js/models/badges.js]: Not so sure.
I can't remove the brackets, namespace and operation are variables.
The parenthesisaround response are part of the arrow function syntax

File: js/models/base.js (hunk 4 above)
Comment (noirbizarre) [inline js/models/base.js]: Thanks :)
I like this one ;)

File: js/models/base.js (hunk 5 above)
Comment (noirbizarre) [inline js/models/base.js]: I think I can, I will ASAP :)

File: js/models/base.js (hunk 6 above)
Comment (noirbizarre) [inline js/models/base.js]: For the arrow function, I will adapt it.
This boolean is very usefull to automatically display a search field or not into datatables.vue.

File: js/models/base.js (hunk 7 above)
Comment (noirbizarre) [inline js/models/base.js]: It smell like an iterator but ... it's not :/
I kept the original method names (next and previous), but I think it iwll be safer to rename it to `nextPage` and `previousPage`

File: js/models/base.js (hunk 8 above)
Comment (noirbizarre) [inline js/models/base.js]: Yes, exactly :)
It need to be implemented

File: js/models/base.js (hunk 9 above)
Comment (noirbizarre) [inline js/models/base.js]: Yep, I will

File: js/models/organization.js (hunk 10 above)
Comment (noirbizarre) [inline js/models/organization.js]: Better than jshint ;)

File: js/models/organization.js (hunk 11 above)
Comment (noirbizarre) [inline js/models/organization.js]: Agreed, can be done later but can be interesting to wrap Swagger into promises

File: js/models/site.js (hunk 12 above)
Comment (noirbizarre) [inline js/models/site.js]: This pattern is only used here et for `Me`.
Data is fetched once on application start.

**Pull request merged**